File: app/src/main/java/at/privatepilotserver/server/websocket.js
const { getClientKey, encrypt, decodeWithKey } = require('./encryption');
const { validateUser } = require('./userAuth')

const fs = require('fs');
const path = require('path');
const sharp = require('sharp');
const WebSocket = require('ws');

const BASE_DIRECTORY = 'C:/Users/lampr/Desktop/fileStorage';


function startWebSocketServer(port) {
    const wss = new WebSocket.Server({ port: port });
    console.log(`WebSocket Server is running on port ${port}`);

    wss.on('connection', async (ws, req) => {
        const encodedName = req.headers['username'];
        const encodedToken = req.headers['authorization'];
        const clientPublicKey = req.headers['publickey'];

        await getClientKey(clientPublicKey);

        const decodedName = decodeWithKey(encodedName);
        const decodedToken = decodeWithKey(encodedToken);

        if (!validateUser(decodedName, decodedToken)) {
            ws.terminate();
            return;
        }

        ws.on('message', async (message) => {
            const buffer = Buffer.from(message);
            const request = buffer.toString();

            if (request.startsWith('GET:')) {
                handleGetRequest(ws, request);
            } else if (request.startsWith('POST:')) {
                handlePostRequest(ws, request);
            } else if (request.startsWith('DELETE:')) {
                handleDeleteRequest(ws, request);
            } else if (request.startsWith('FIND:')) {
                handleSearchRequest(ws, request);
            } else if (request.startsWith('UPDATE:')) {
                handleUpdateRequest(ws, request);
            }
        });

        ws.on('close', () => {
            console.log('Client disconnected');
        });
    });
}

let pointer = BASE_DIRECTORY;

async function sendToClient(ws, message) {
    const encryptedMessage = await encrypt(message);

    ws.send(encryptedMessage, (error) => {
        if (error) {
            console.error(`Error sending file: ${error.message}`);
        }
    });
}



function handleGetRequest(ws, request) {
    const folderOrFilePath = request.slice('GET:'.length);
    const fullPath = path.join(BASE_DIRECTORY, folderOrFilePath);

    if (fs.existsSync(fullPath)) {
        if (fs.statSync(fullPath).isDirectory()) {
            const items = listFileMetadata(fullPath);
            const itemsJSON = JSON.stringify({'items': items });
            pointer = fullPath;
            const relative_path = path.relative(BASE_DIRECTORY, pointer)

            console.log(`Listing files in: ${fullPath}`);

            sendToClient(ws, `/${relative_path};${itemsJSON}`);
            sendFilePreviews(ws, items);
        } else if (fs.statSync(fullPath).isFile()) {
            console.log(`Sending file: ${fullPath}`);
            sendFileToClient(ws, fullPath);
        }
    }
}

function handlePostRequest(ws, request) {
    const relativeFilePath = request.slice('POST:'.length);
    const relativeDirectory = path.dirname(relativeFilePath);

    const folderPath = path.join(BASE_DIRECTORY, relativeDirectory);
    const filePath = path.join(BASE_DIRECTORY, relativeFilePath);

    if (!fs.existsSync(folderPath)) {
        fs.mkdirSync(folderPath, { recursive: true });
    }

    ws.once('message', (fileData) => {
        console.log(`Received file data for: ${filePath}`);
        saveFileToServer(filePath, fileData);
    });
}

function handleDeleteRequest(ws, request) {
    const folderOrFilePath = request.slice('DELETE:'.length);
    const fullPath = path.join(BASE_DIRECTORY, folderOrFilePath);

    if (fs.existsSync(fullPath)) {
        fs.unlinkSync(fullPath);
        console.log(`Deleted file: ${fullPath}`);
    }
}

function listFileMetadata(fullPath) {
    const items = [];
    const itemsInDirectory = fs.readdirSync(fullPath);

    for (const item of itemsInDirectory) {
        const itemPath = path.join(fullPath, item);
        const relativePath = path.relative(BASE_DIRECTORY, itemPath).replace(/\\/g, '/');
        const itemStat = fs.statSync(itemPath);
        const itemType = itemStat.isFile() ? 'file' : 'folder';

        items.push({
            name: item,
            path: relativePath,
            type: itemType,
            size: itemStat.isFile() ? itemStat.size : null,
            last_modified: new Date(itemStat.mtime).getTime() / 1000
        });
    }

    return items;
}

function sendFileToClient(ws, filePath) {
    fs.readFile(filePath, (err, fileData) => {
        if (err) {
            console.error(`Error reading file: ${err.message}`);
        } else {
            console.log(`Sending file data for: ${filePath}`);
            ws.send(fileData); //quickfix to send files to client, currently unencrypted
        }
    });
}

function saveFileToServer(filePath, fileData) {
    console.log(`Saving file to: ${filePath}`);
    fs.writeFile(filePath, fileData, (err) => {
        if (err) {
            console.error(`Error saving file: ${err.message}`);
        } else {
            console.log(`Saved file to: ${filePath}`);
        }
    });
}

function handleSearchRequest(ws, request) {
    const searchQuery = request.slice('FIND:'.length);
    const searchRegex = new RegExp(searchQuery, 'i');
    const searchResults = searchFiles(pointer, searchRegex);
    const resultsJSON = JSON.stringify({ 'items': searchResults });

    const relative_path = path.relative(BASE_DIRECTORY, pointer)

    console.log(`Search results for query "${searchQuery}" in ${pointer}:`);
    sendToClient(ws, `/${relative_path};${resultsJSON}`);
}

function searchFiles(fullPath, searchRegex) {
    const items = [];
    const itemsInDirectory = fs.readdirSync(fullPath);

    for (const item of itemsInDirectory) {
        const itemPath = path.join(fullPath, item);
        const itemStat = fs.statSync(itemPath);
        if (searchRegex.test(item)) {
            const relativePath = path.relative(BASE_DIRECTORY, itemPath).replace(/\\/g, '/');
            const itemType = itemStat.isFile() ? 'file' : 'folder';

            items.push({
                name: item,
                path: relativePath,
                type: itemType,
                size: itemStat.isFile() ? itemStat.size : null,
                last_modified: new Date(itemStat.mtime).getTime() / 1000
            });
        }

        if (itemStat.isDirectory()) {
            const subdirectoryResults = searchFiles(itemPath, searchRegex);
            items.push(...subdirectoryResults);
        }
    }

    return items;
}

function handleUpdateRequest(ws, request) {
    const [sourcePath, destinationPath] = request.slice('UPDATE:'.length).split(';');
    const fullSourcePath = path.join(BASE_DIRECTORY, sourcePath);
    const fullDestinationPath = path.join(BASE_DIRECTORY, destinationPath);

    console.log(`Source Path: ${fullSourcePath}`);
    console.log(`Destination Path: ${fullDestinationPath}`);

    try {
        if (fs.existsSync(fullSourcePath)) {
            const isSourceDirectory = fs.lstatSync(fullSourcePath).isDirectory();
            let finalDestinationPath;

            if (isSourceDirectory) {
                finalDestinationPath = path.join(fullDestinationPath, path.basename(fullSourcePath));
            } else {
                finalDestinationPath = path.join(fullDestinationPath, path.basename(fullSourcePath));
            }

            console.log(`Final Destination Path: ${finalDestinationPath}`);
            fs.renameSync(fullSourcePath, finalDestinationPath);
            console.log(`Moved file/directory from ${fullSourcePath} to ${finalDestinationPath}`);
        } else {
            console.error(`Source path does not exist: ${fullSourcePath}`);
        }
    } catch (error) {
        console.error(`Error moving file/directory: ${error.message}`);
    }
}

async function sendFilePreviews(ws, items) {
    for (const item of items) {
        const filePath = path.join(BASE_DIRECTORY, item.path);
        const fileExtension = path.extname(filePath).toLowerCase();
        let preview;

        if (/\.(jpg|png|gif)$/i.test(fileExtension)) {
            preview = await generateImagePreview(filePath);
        }

        if (preview != null) {
            console.log(`Sending preview for file: ${item.name}`);
            sendToClient(ws, `base64;${item.path};${preview}`);
        }
    }
}

async function generateImagePreview(filePath) {
    try {
        const imageBuffer = fs.readFileSync(filePath);
        const resizedBuffer = await sharp(imageBuffer)
            .resize({ width: 120, height: 150 })
            .toBuffer();

        const base64Preview = resizedBuffer.toString('base64');
        return base64Preview;
    } catch (error) {
        console.error(`Error generating preview for ${filePath}: ${error.message}`);
        return null;
    }
}

module.exports = {
    startWebSocketServer,
};

File: app/src/main/java/at/privatepilot/server/Websocket.kt
package at.privatepilot.server

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.ThumbnailUtils
import at.privatepilot.server.model.Metadata
import at.privatepilot.server.model.MetadataResponse
import com.google.gson.Gson
import io.ktor.application.install
import io.ktor.http.cio.websocket.DefaultWebSocketSession
import io.ktor.http.cio.websocket.Frame
import io.ktor.http.cio.websocket.readText
import io.ktor.routing.routing
import io.ktor.server.engine.embeddedServer
import io.ktor.server.netty.Netty
import io.ktor.websocket.webSocket
import kotlinx.coroutines.channels.consumeEach
import okio.ByteString.Companion.toByteString
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.file.Paths
import java.util.Base64


class Websocket(private val port: Int, private val encryption: Encryption, private val context: Context) {
    private val BASE_DIRECTORY = "server_"
    private val home = context.getExternalFilesDir(BASE_DIRECTORY)
    private var pointer = BASE_DIRECTORY

    private val controller = Controller(context)

    init {
        val userAuth = UserAuth(context)
        embeddedServer(Netty, port) {
            install(io.ktor.websocket.WebSockets)
            routing {
                webSocket("/ws") {
                    controller.fileExist(BASE_DIRECTORY)
                    val username = call.request.headers["username"]
                    val authorization = call.request.headers["authorization"]
                    val publickey = call.request.headers["publickey"]
                    encryption.getPublicKeyFromString(publickey)
                    val decodedUsername = username?.let { encryption.decrypt(it) }
                    val decodedauthorization = authorization?.let { encryption.decrypt(it) }
                    userAuth.validateUser(decodedUsername, decodedauthorization)

                    incoming.consumeEach { frame ->
                        when (frame) {
                            is Frame.Text -> {
                                var decryptedMessage = frame.readText()

                                if (decryptedMessage != "Ping") {
                                    decryptedMessage = encryption.decrypt(decryptedMessage)

                                    when {
                                        decryptedMessage.startsWith("GET:") -> handleGetRequest(this, decryptedMessage.substring(4))
                                        decryptedMessage.startsWith("POST:") -> handlePostRequest(this, decryptedMessage.substring(5))
                                        decryptedMessage.startsWith("DELETE:") -> handleDeleteRequest(this,decryptedMessage.substring(8))
                                        decryptedMessage.startsWith("FIND:") -> handleSearchRequest(this, decryptedMessage.substring(5))
                                        decryptedMessage.startsWith("UPDATE:") -> handleUpdateRequest(this, decryptedMessage.substring(7))
                                    }
                                }
                            }
                            // Handle other types of frames if needed
                            else -> {outgoing.send(Frame.Text("toast; request not recognized"))}
                        }
                    }
                }
            }
        }.start(wait = true)
    }

    private suspend fun sendText(session: DefaultWebSocketSession, message: String) {
        try {
            val encryptedMessage = encryption.encrypt(message)
            session.outgoing.send(Frame.Text(encryptedMessage))
            println("Sent message: $message")
        } catch (e: Exception) {
            println("Error sending WebSocket message: ${e.message}")
        }
    }

    private suspend fun sendFileToClient(session: DefaultWebSocketSession, filePath: String) {
        try {
            val fileData = File(filePath).readBytes().toByteString()
            val encryptedFile = encryption.encrypt(fileData)
            session.outgoing.send(Frame.Text(encryptedFile))
            println("Sent file: $filePath")
        } catch (e: Exception) {
            println("Error sending file: ${e.message}")
        }
    }

    private suspend fun handleGetRequest(session: DefaultWebSocketSession, request: String) {
        try {
            val requestedFile = controller.fileExist(request)

            if (requestedFile.exists()) {
                if (requestedFile.isDirectory) {

                    pointer = home?.let { requestedFile.relativeTo(it).path }.toString()

                    val items = listFileMetadata(requestedFile)
                    // Send the response with file metadata
                    sendText(session, "$request;${Gson().toJson(items)}")
                    sendFilePreviews(session, items)
                } else if (requestedFile.isFile) {
                    // Print a message for debugging
                    println("Sending file: $requestedFile")

                    // Send the file to the client
                    sendFileToClient(session, requestedFile.absolutePath)
                }
            } else {
                // Print a message for debugging
                println("Resource not found: $requestedFile")

                // Send a response indicating that the resource was not found
                sendText(session, "resource not found")
            }
        } catch (e: Exception) {
            // Handle exceptions, log an error, and inform the client about the error
            println("Error handling GET request: ${e.message}")
            sendText(session, "error handling GET request: ${e.message}")
        }
    }

    private suspend fun handlePostRequest(session: DefaultWebSocketSession, request: String) {

        val newFile = controller.fileExist(request, true)

        if (!newFile.isDirectory) {
            val frame = session.incoming.receive()
            if (frame is Frame.Text) {
                val message = frame.readText()
                val fileData = encryption.decryptFile(message)
                saveFileToServer(newFile, fileData.toByteArray(), session)
            }
            return
        } else handleGetRequest(session, "GET:${pointer}")
    }

    private suspend fun handleDeleteRequest(session: DefaultWebSocketSession, request: String) {
        val file = context.getExternalFilesDir("${BASE_DIRECTORY}/${request}")

        file?.delete()

        println("Deleted file: $request")
        handleGetRequest(session, pointer)
    }

    private fun listFileMetadata(directory: File): MetadataResponse {
        val items = mutableListOf<Metadata>()

        directory.listFiles()?.forEach { item ->
            val relativePath = home?.let { item.relativeTo(it).path }
            val itemType = if (item.isFile) "file" else "folder"

            val metadata = Metadata(item.name,
                relativePath ?: "",
                itemType,
                if (item.isFile) item.length().toInt() else 0,
                item.lastModified() / 1000.0
            )

            items.add(metadata)
        }

        return MetadataResponse(items)
    }

    private suspend fun saveFileToServer(file: File, fileData: ByteArray, session: DefaultWebSocketSession) {
        try {
            file.writeBytes(fileData)
            handleGetRequest(session, "GET:${pointer}")
        } catch (e: Exception) {
            println("Error saving file: ${e.message}")
        }
    }

    private suspend fun handleSearchRequest(session: DefaultWebSocketSession, request: String) {
        val searchQuery = request
        val searchRegex = Regex(searchQuery, RegexOption.IGNORE_CASE)

        val pointerDirectory = if (pointer.isNullOrBlank()) home else context.getExternalFilesDir("${BASE_DIRECTORY}/${pointer}")

        val searchResults = pointerDirectory?.let { searchFiles(it, searchRegex) }

        if (searchResults != null) {
            sendText(session, "$request;${Gson().toJson(searchResults)}")
            sendFilePreviews(session, searchResults)
        }
    }

    private fun searchFiles(searchFolder: File, searchRegex: Regex): MetadataResponse {
        val items = mutableListOf<Metadata>()

        searchFolder.listFiles()?.forEach { item ->
            val relativePath = home?.let { item.relativeTo(it).path }
            val itemType = if (item.isFile) "file" else "folder"

            if (searchRegex.containsMatchIn(item.name)) {
                val metadata = Metadata( item.name, relativePath ?: "", itemType, if (item.isFile) item.length().toInt() else 0, item.lastModified() / 1000.0 )

                items.add(metadata)
            }

            if (item.isDirectory) {
                val subdirectoryResults = searchFiles(item, searchRegex)
                items.addAll(subdirectoryResults.items)
            }
        }

        return MetadataResponse(items)
    }

    private suspend fun handleUpdateRequest(session: DefaultWebSocketSession, request: String) {
        val (sourcePath, destinationPath) = request.split(";")
        val sourceFile = context.getExternalFilesDir("${BASE_DIRECTORY}${sourcePath}")

        try {
            if (sourceFile?.exists() == true) {
                val destinationFile = controller.fileExist(destinationPath, true)
                sourceFile.copyTo(destinationFile, true)
                if (destinationFile.exists())
                    sourceFile.delete()
            } else {
                println("Source path does not exist: $sourcePath")
            }
        } catch (error: Exception) {
            println("Error moving file/directory: ${error.message}")
        }
        handleGetRequest(session, pointer)
    }

    private suspend fun sendFilePreviews(session: DefaultWebSocketSession, items: MetadataResponse) {
        for (item in items.items) {
            val filePath = Paths.get(BASE_DIRECTORY, item.path.toString()).toString()
            val fileExtension = File(filePath).extension
            var preview: String? = null

            if (fileExtension.matches(Regex("(jpg|png|gif)$", RegexOption.IGNORE_CASE))) {
                preview = generateImagePreview(filePath)
            }

            if (preview != null) {
                println("Sending preview for file: ${item.name}")
                sendText(session, "base64;${item.path};$preview")
            }
        }
    }

    private fun generateImagePreview(filePath: String): String? {
        try {
            val imageFile = context.getExternalFilesDir(filePath)
            if (imageFile != null) {
                val imageThumbnail = ThumbnailUtils.extractThumbnail(
                    BitmapFactory.decodeFile(imageFile.path),
                    120, 150
                )

                val byteArrayOutputStream = ByteArrayOutputStream()
                imageThumbnail.compress(Bitmap.CompressFormat.PNG, 100, byteArrayOutputStream)
                val byteArray = byteArrayOutputStream.toByteArray()

                return Base64.getEncoder().encodeToString(byteArray)
            }
        } catch (error: Exception) {
            println("Error generating preview for $filePath: ${error.message}")
        }
        return null
    }
}
